SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -373,12 +374,22 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
 
 local infix:20 " ≈ʷ " => EqualUpTo w
 
--- TODO: These sorries are difficult, and will be proven in a later Pull Request.
+-- TODO: This sorry is difficult, and will be proven in a later Pull Request.
 theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
   sorry
 
```

Comment:
By the way, I believe this one should not actually be that hard either. Similar to `neg`, we can in fact reduce this fairly easily to known facts:
```lean
theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
  have : x - y = x + ~~~y + 1#w := by
    simp only [BitVec.sub_toAdd, BitVec.neg_eq_not_add, BitVec.ofNat_eq_ofNat]
    ac_rfl
  calc
  _ ≈ʷ ofBitVec (x + ~~~y + 1#w)            := by rw [this]
  _ ≈ʷ ofBitVec _ + ofBitVec _              := ofBitVec_add
  ...
```

I tried stitching the proof together and failed, but I believe @AtticusKuhn might know how to get the proper congruences to make this all go through (it sure would be nice to have generalized rewriting!)

Reply:
You are correct, proving "ofBitVec_sub" is not very difficult:

```lean
theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
  calc
  _ ≈ʷ (@ofBitVec w (x + (- y))) := by rw [BitVec.sub_add_neg]
  _ ≈ʷ ((@ofBitVec w x) + (@ofBitVec w (-y))) := ofBitVec_add
  _ ≈ʷ ((@ofBitVec w x) + (- @ofBitVec w (y))) := add_congr equal_up_to_refl ofBitVec_neg
  _ ≈ʷ (ofBitVec x) - (ofBitVec (y)) := by rw [← sub_add_neg]
```

However, I find proving "a - b = a + (-b) to be significantly more difficult:


```lean
theorem sub_add_neg' {a b : BitStream} : a - b = a + (-b) := by
  have sub_add_lemma (i : Nat) :
      let y := b.negAux
      let x := a.addAux (fun n => (y n).1) i
      a.subAux b i = ⟨x.2, !(y i).2 &&  !x.1 ⟩  := by
    induction' i with i ih
    · simp [subAux,addAux,negAux, BitVec.adcb]
      cases (b 0 ) ;
      · simp
      · simp
    · simp [subAux,addAux,negAux, BitVec.adcb]
      rw [ih]
      simp
      cases a (i + 1)
      all_goals cases b (i + 1)
      all_goals simp
      all_goals constructor
      all_goals sorry
  ext i
  simp only [HAdd.hAdd, HSub.hSub, Neg.neg, Sub.sub, BitStream.sub,Add.add, BitStream.add]
  unfold neg
  simp [sub_add_lemma i]
```

Reply:
Could you push a draft PR such that we see where things become difficult?

Reply:
> Could you push a draft PR such that we see where things become difficult?

That sounds like a good idea, I've done that here:
https://github.com/opencompl/lean-mlir/pull/554/files